### PR TITLE
Allow proxy to path

### PR DESCRIPTION
**Pull request by jason-jackson**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -28,6 +30,33 @@ func init() {
 	httpcaddyfile.RegisterHandlerDirective("dynamic_proxy", parseCaddyfile)
 }
 
+type AuthSite struct {
+	To   string
+	Path string
+}
+
+func NewAuthSite(site string) (AuthSite, error) {
```
Comment (Schparky) [inline main.go:38]: My feeling is that a `New____` function (a/k/a Contstructor) shouldn't return an error, though I don't have a source to back me up on that. If you would like to try this, I recommend splitting this into two functions:
```go
func NewAuthSite(siteURL *url.URL) *AuthSite {
	result := AuthSite{
		To:   siteURL.Host,
		Path: siteURL.Path,
	}
	if siteURL.Port() == "" {
		result.To += ":80"
	}
	return result
}
```
and
```go
func toURL(addr string) (*url.URL, error) {
  	if m, _ := regexp.MatchString("^\\w+://", addr); !m {
		site = "http://" + site
	}
	return url.Parse(addr)
}
```

**Review by Schparky — APPROVED** [review]

File: main_test.go [hunk 2]
```diff
@@ -11,19 +11,81 @@ import (
...
+			want:    AuthSite{"http.com:9000", ""},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			got, err := NewAuthSite(tc.value)
+			if tc.wantErr {
+				assert.Error(err)
+				assert.Contains(err.Error(), tc.err)
```
Comment (Baggerone) [inline main_test.go:66]: It's not much of a difference, but there is assert.ErrorContains()

**Review by Baggerone — APPROVED** [review]

File: main.go [hunk 3]
```diff
@@ -75,10 +105,23 @@ func (p *Proxy) Provision(ctx caddy.Context) error {
 	}
 	p.auth.TokenSecret = string(secret)
 
-	p.sites = make(map[string]string)
-	p.sites[p.auth.SiteOneLevel] = p.auth.SiteOne
-	p.sites[p.auth.SiteTwoLevel] = p.auth.SiteTwo
-	p.sites[p.auth.SiteThreeLevel] = p.auth.SiteThree
+	p.sites = make(map[string]AuthSite)
+	p.sites[p.auth.SiteOneLevel], err = NewAuthSite(p.auth.SiteOne)
```
[the remark this answers is not in the document]
Comment (jason-jackson) [reply main.go:109]: This was done mainly because of an [issue](https://github.com/kelseyhightower/envconfig/issues/180) with kelseyhightower/envconfig where he use Split instead of SplitN or Cut. So urls with ports cause errors. Given that the project hasn't been updated in a few years, I'm not holding my breath for that fix.

I could make a custom decoder and use that instead, it probably won't actually be much more work. Given that AuthSite is pretty much a custom decoder, I think it's worth doing.

File: main.go [hunk 4]
```diff
@@ -130,21 +173,27 @@ func (p Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request, next caddyhttp.
...
 
-	to, err := p.authRedirect(r)
+	upstream, err := p.authRedirect(r)
 	if err != nil {
 		return err
 	}
-	caddyhttp.SetVar(r.Context(), "upstream", to)
-	p.log.Info("setting upstream to " + to)
+
+	if upstream == p.api || upstream.Path == "" {
```
Comment (forevermatt) [inline main.go:181]: I don't quite understand this if-condition. It looks like it's saying, "If we're sending them to the authorization API, or if the upstream server we're sending them to doesn't define a specific URL path, then use whatever URL path they requested."

However, if we're sending them to the authorization API, then we only want them to go to a single, pre-defined URL path... the one that sets the JWT with the authorization data.

Would you mind clarifying this condition's purpose for me? I suspect I'm still missing/misunderstanding something about it.

File: main.go [hunk 5]
```diff
@@ -130,21 +173,27 @@ func (p Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request, next caddyhttp.
...
 	return next.ServeHTTP(w, r)
 }
 
-func (p Proxy) authRedirect(r *http.Request) (string, error) {
+func (p Proxy) authRedirect(r *http.Request) (AuthSite, error) {
 	// if no cookie, redirect to get new cookie
 	cookie, err := r.Cookie(p.auth.CookieName)
 	if err != nil {
-		return p.ManagementAPI, nil
+		return p.api, nil
```
Comment (forevermatt) [inline main.go:196]: Out of curiosity, why the change from "ManagementAPI" to "api"?

File: main.go [hunk 6]
```diff
@@ -28,6 +30,33 @@ func init() {
 	httpcaddyfile.RegisterHandlerDirective("dynamic_proxy", parseCaddyfile)
 }
 
+type AuthSite struct {
```
[the remark this answers is not in the document]
Comment (jason-jackson) [reply main.go:33]: Changing it to `package main` causes caddy to throw errors.

File: main_test.go (hunk 2 above)
Comment (jason-jackson) [reply main_test.go:66]: Thanks, this does help it fail tests better (if it fails here).

File: main.go (hunk 5 above)
Comment (jason-jackson) [reply main.go:209]: mainly needed so that I could get around the reverse_proxy and paths. It's quite possible I can remove this if/when we switch to redirect instead of reverse proxy to management api.

File: main.go (hunk 4 above)
Comment (jason-jackson) [reply main.go:194]: We want to redirect them back to wherever they were before, not just "/", though I'll probably need to adjust it some to use a query param (not sure if I can rewrite with a query param included or not, it didn't look like it when I tried at first, but I may have not been doing it correctly).

File: main_test.go [hunk 7]
```diff
@@ -11,19 +11,81 @@ import (
...
+		{
+			name:    "ftp",
+			value:   "ftp://ftp:9000",
+			wantErr: false,
+			want:    AuthSite{"ftp:9000", ""},
+		},
+		{
+			name:    "http",
+			value:   "http://http:9000",
+			wantErr: false,
```
[the remark this answers is not in the document]
Comment (jason-jackson) [reply main_test.go:50]: Added an error case.

File: main.go (hunk 1 above)
Comment (jason-jackson) [reply main.go]: No longer have a `New___` function now parsing directly from CLI

File: local-example.env [hunk 8]
```diff
@@ -1,12 +1,6 @@
 # For proxy JWT (required)
 AUTH_COOKIE_NAME=
 AUTH_TOKEN_SECRET=
+AUTH_SITES=
```
Comment (Baggerone) [inline local-example.env:4]: Maybe add a comment here with an example.

File: main.go [hunk 9]
```diff
@@ -30,42 +31,69 @@ func init() {
 	httpcaddyfile.RegisterHandlerDirective("dynamic_proxy", parseCaddyfile)
 }
 
+type AuthSites map[string]AuthSite
+
+func (a *AuthSites) Decode(input string) error {
+	*a = make(AuthSites)
+	sites := strings.Split(input, ";")
+	for _, site := range sites {
```
Comment (Baggerone) [inline main.go:39]: Maybe make this `for _, s := ... `
Or, if you prefer, in line 44 change the first `site` to something else, so that you don't assign a new value to the `for` variable.

File: main.go [hunk 10]
```diff
@@ -30,42 +31,69 @@ func init() {
...
-		return AuthSite{}, err
+		return err
 	}
 
-	result := AuthSite{
-		To:   u.Host,
-		Path: u.Path,
+	// Add protocol if it is missing (needed for url parse)
+	if !re.MatchString(input) {
+		input = "http://" + input
```
Comment (Baggerone) [inline main.go:77]: Is this safe enough or would it be better just to return an error and say that the protocol is required?
I'm just wondering if in situations where it should really be `https` that the resulting problem/failure will be hidden from us.

**Review by Baggerone — APPROVED** [review]

File: main.go (hunk 10 above)
Comment (jason-jackson) [reply main.go:77]: The protocol is currently not used, it is just there so url.Parse can parse it. I just get the host and path.
